com.unity.ml-agents.extensions/Runtime/Match3/AbstractBoard.cs

```diff
@@ -0,0 +1,232 @@
+        /// <returns></returns>
+        public IEnumerable<Move> AllMoves()
+        {
+            var currentMove = Move.FromMoveIndex(0, Rows, Columns);
+            for (var i = 0; i < NumMoves(); i++)
+            {
+                yield return currentMove;
+                currentMove.Advance(Rows, Columns);
+            }
+        }
+
+        /// <summary>
+        /// Iterate through all valid Moves on the board.
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerable<Move> ValidMoves()
+        {
+            var currentMove = Move.FromMoveIndex(0, Rows, Columns);
+            for (var i = 0; i < NumMoves(); i++)
+            {
+                if (IsMoveValid(currentMove))
+                {
+                    yield return currentMove;
+                }
+                currentMove.Advance(Rows, Columns);
+            }
+        }
+
+        /// <summary>
+        /// Iterate through all invalid Moves on the board.
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerable<Move> InvalidMoves()
+        {
+            var currentMove = Move.FromMoveIndex(0, Rows, Columns);
+            for (var i = 0; i < NumMoves(); i++)
+            {
+                if (!IsMoveValid(currentMove))
+                {
+                    yield return currentMove;
+                }
+                currentMove.Advance(Rows, Columns);
+            }
+        }
+
+        /// <summary>
+        /// Returns true if swapped the cells specified by the move would result in
+        /// 3 or more cells of the same type in a row. This assumes that all pieces are allowed
+        /// to be moved; to add extra logic, incorporate it into you IsMoveValid() method.
+        /// </summary>
+        /// <param name="move"></param>
+        /// <returns></returns>
+        public bool SimpleIsMoveValid(Move move)
+        {
+            using (TimerStack.Instance.Scoped("SimpleIsMoveValid"))
+            {
+                var moveVal = GetCellType(move.Row, move.Column);
+                var (otherRow, otherCol) = move.OtherCell();
+                var oppositeVal = GetCellType(otherRow, otherCol);
+
+                // Simple check - if the values are the same, don't match
+                // This might not be valid for all games
+                {
+                    if (moveVal == oppositeVal)
+                    {
+                        return false;
+                    }
+                }
+
+                bool moveMatches = CheckHalfMove(otherRow, otherCol, moveVal, move.Direction);
+                if (moveMatches)
+                {
+                    // early out
+                    return true;
+                }
+
+                bool otherMatches = CheckHalfMove(move.Row, move.Column, oppositeVal, move.OtherDirection());
+                return otherMatches;
+            }
+        }
+
+        /// <summary>
+        /// Check if the "half" of a move is matches 3 or more.
```

Comment:
```suggestion
        /// Check if the "half" of a move matches 3 or more.
```

Reply:
What is a half move ?

Reply:
Updated:
```
        /// Check if one of the cells that is swapped during a move matches 3 or more.
        /// Since these checks are similar for each cell, we consider the Move as two "half moves".
```